core/templates/pages/exploration-player-page/services/question-player-engine.service.spec.ts

```diff
@@ -0,0 +1,873 @@
+      .and.returnValue(Promise.resolve(exploration));
+    spyOn(answerClassificationService, 'getMatchingClassificationResult')
+      .and.returnValue(answerClassificationResult);
+    spyOn(expressionInterpolationService, 'processHtml')
+      .and.callFake((html, envs) => html);
+
+    questionPlayerEngineService.init(
+      multipleQuestionsBackendDict, successHandler, failHandler);
+    let languageCode = questionPlayerEngineService.getLanguageCode();
+
+    expect(languageCode).toBe('en');
+
+    questionPlayerEngineService.submitAnswer(
+      answer, interactionRulesService, successCallback);
+    questionPlayerEngineService.recordNewCardAdded();
+
+    languageCode = questionPlayerEngineService.getLanguageCode();
+    expect(languageCode).toBe('br');
+
+    questionPlayerEngineService.submitAnswer(
+      answer, interactionRulesService, successCallback);
+    questionPlayerEngineService.recordNewCardAdded();
+
+    languageCode = questionPlayerEngineService.getLanguageCode();
+    expect(languageCode).toBe('ab');
+  });
+
+  it('should always return false when calling \'isInPreviewMode()\'', () => {
+    let previewMode = questionPlayerEngineService.isInPreviewMode();
+
+    expect(previewMode).toBe(false);
+  });
+
+  it('should show warning message while loading a question ' +
+    'if the question html content is null', () => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+
+    singleQuestionBackendDict.question_state_data
+      .content.html = null;
+    let alertsServiceSpy =
+      spyOn(alertsService, 'addWarning').and.returnValue();
+    spyOn(expressionInterpolationService, 'processHtml')
+      .and.callFake((html, envs) => html);
+
+    questionPlayerEngineService.init(
+      [singleQuestionBackendDict], successHandler, failHandler);
+
+    expect(alertsServiceSpy).toHaveBeenCalledWith('Expression parsing error.');
+  });
+
+  it('should not load questions if there are no questions', () => {
+    let successHandler = jasmine.createSpy('success');
+    let failHandler = jasmine.createSpy('fail');
+
+    questionPlayerEngineService.init(
+      [], successHandler, failHandler);
+    expect(successHandler).not.toHaveBeenCalled();
+    expect(failHandler).toHaveBeenCalled();
+  });
+
+  describe('on submitting answer ', () => {
+    it('should submit answer correctly', () => {
+      let successCallback = jasmine.createSpy('success');
+      let successHandler = jasmine.createSpy('success');
+      let failHandler = jasmine.createSpy('fail');
+      let answer = 'answer';
+      let interactionRulesService: InteractionRulesService;
+      let answerClassificationResult = new AnswerClassificationResult(
+        outcomeObjectFactory
+          .createNew('default', '', '', []), 1, 0, 'default_outcome'
+      );
+      answerClassificationResult.outcome.labelledAsCorrect = true;
+
+
+      spyOn(answerClassificationService, 'getMatchingClassificationResult')
```

Comment:
Never have two consecutive newlines, here and elsewhere. Just one suffices.

Reply:
done